src/psyclone/psyir/nodes/routine.py

```diff
@@ -51,9 +53,14 @@ class Routine(Schedule, CommentableMixin):
     A sub-class of a Schedule that represents a subroutine, function or
     program unit.
 
-    :param str name: the name of this routine.
-    :param bool is_program: whether this Routine represents the entry point \
-                            into a program (e.g. Fortran Program or C main()).
+    :param symbol: the Symbol used to represent this Routine in its Container.
+                   If it is not supplied a RoutineSymbol will be created with
+                   default arguments instead.
```

Comment:
The reason we're not fine with it if its already there is that we have no way of determining why its already there - e.g. is there another Routine in this scope using that symbol that we've piggy-backed on (which would be an error). I'll add a note for this.

Reply:
I've written a note to attempt to emphasize the why, and removed the out-of-date section of the docstring I think.

Reply:
> e.g. is there another Routine in this scope using that symbol that we've piggy-backed on

@LonelyCat124 but wouldn't this be another RoutineSymbol with the same name, would it be fine to let it pass if it is the same object ("is" comparison)?

Reply:
Ah you mean like attaching the same routine node (or a copy of it) 2 times? Since we have the scope, could a scope.walk(Routine) tell us if thats the case and then complain, but be fine if not found?

Reply:
I think mostly I'm wary of it to avoid mistakes, e.g. if you have this code

```
subroutine my_sub()
...
end subroutine my_sub
```

and you did
```
fc = my_sub.ancestor(FileContainer)
routine_symbol = fc.symbol_table.lookup("my_sub")
# Meant to do .replace but made a mistake
replacement = Routine(routine_symbol, parent=fc)
```

I think this should fail, because otherwise PSyclone may make a Fortran output like:
```
subroutine my_sub()
...
end subroutine

subroutine my_sub()
...
end subroutine
```

which I think is unlikely to be intended output.

I guess your solution could work (though it would need to be `scope.walk(Routine, CodeBlock)` and be able to check the CodeBlock name as well (I'm not sure how this works). I can change it to that if you think it'd be appropriate.

Reply:
Added support for this, just need to update the documentation.

Reply:
Also refactored the code slightly.